### DISCUSSION TITLE
Delay in parallel control of multiple crazyflies using 2 crazyradios

### DISCUSSION THREAD

**Discussion by adarsh2798** [bitcraze/discussions#1181] [Q&A]
Hi, I am new to crazyflies. Please excuse me if this is a trivial question

I am using 5 crazyflies and I am using cflib and have written a simple python script where all 5 crazyfleis must takeoff and hover at some point and land back. This process must happen simulatenously for all crazyflies. I am using 2 crazyradios and assigned 3 and 2 crazyfleis per radio. My uris are:
```
URI1='radio://0/80/2M/E7E7E7E701'
URI2='radio://0/80/2M/E7E7E7E702'
URI3='radio://0/80/2M/E7E7E7E703'
URI4='radio://1/100/2M/E7E7E7E704'
URI5='radio://1/100/2M/E7E7E7E705'

uris={URI1,URI2,URI3,URI4,URI5}
```

then I am using swarm class parallel method :
```
if __name__=="__main__":
        cflib.crtp.init_drivers()
        factory=CachedCfFactory(rw_cache="./cache")
        with Swarm(uris,factory=factory) as swarm:
     
          swarm.parallel(get_data)
          time.sleep(2)
          swarm.parallel(test)
```
        
  But probelm i face is here swarm.parallel() gets executed first for URI4,5. Then after those 2 have finished, URIs 1,2,3 start. Why this delay?? Shouldnt swarm ensure parallel action for all crazyflies by opening a separate thread for each? Or it works only for single crazyradio?
 
below is my full code:

```




import time
import csv
import datetime
import numpy as np
import cflib.crtp
from cflib.crazyflie.log import LogConfig
from cflib.crazyflie.swarm import CachedCfFactory
from cflib.crazyflie.swarm import Swarm
from cflib.crazyflie.syncLogger import SyncLogger
from cflib.positioning.motion_commander import MotionCommander
from cflib.crazyflie.commander import Commander
import logging
from cflib.crazyflie.log import LogConfig
from cflib.crazyflie.syncLogger import SyncLogger


import matplotlib.pyplot as plt


URI1='radio://0/80/2M/E7E7E7E701'
URI2='radio://0/80/2M/E7E7E7E702'
URI3='radio://0/80/2M/E7E7E7E703'
URI4='radio://1/20/2M/E7E7E7E704'
URI5='radio://1/20/2M/E7E7E7E705'

uris={URI1,URI2,URI3,URI4,URI5}


def get_R(yaw):
     yaw=yaw*np.pi/180
     R= np.array([[np.cos(yaw), -np.sin(yaw)], 
                         [np.sin(yaw),  np.cos(yaw)]])
     return R


def log_callback(timestamp,data,logconf):
    
   
       pos_x=data['stateEstimate.x']
       pos_y=data['stateEstimate.y']
       pos_z=data['stateEstimate.z']
       
       roll=data['stateEstimate.roll']
       pitch=data['stateEstimate.pitch']
       yaw=data['stateEstimate.yaw']
       
       cf_id=logconf.uri[-2:]
       
      


       print("quad ", cf_id,":")
       #print("pos_xyz:",pos_x,pos_y,pos_z)
       #print("ori_rpy: ",roll,pitch,yaw)


def get_data(scf):
    logconf = LogConfig(name='Position', period_in_ms=100)
    logconf.add_variable('stateEstimate.x', 'float')
    logconf.add_variable('stateEstimate.y', 'float')
    logconf.add_variable('stateEstimate.z', 'float')
    logconf.add_variable('stateEstimate.roll','float')
    logconf.add_variable('stateEstimate.pitch','float')
    logconf.add_variable('stateEstimate.yaw','float')
    logconf.uri=scf.cf.link_uri
    scf.cf.log.add_config(logconf)
    
    logconf.data_received_cb.add_callback(log_callback)
    
    logconf.start()
    




       
def test(scf):
    
    v=np.zeros((2,1))
    cf=scf.cf
    cf_id=cf.link_uri[-2:]
    cf_id=int(cf_id[-1])
    with MotionCommander(scf,default_height=0.5) as mc:
      time.sleep(2)
      
      mc.stop()
      

          
          
    

if __name__=="__main__":
    cflib.crtp.init_drivers()
    factory=CachedCfFactory(rw_cache="./cache")
    with Swarm(uris,factory=factory) as swarm:
     
       swarm.parallel(get_data)
       time.sleep(2)
       swarm.parallel(test)
  





```

**Comment by gemenerik** [comment]
I cannot reproduce your problem. What version of Python, cflib, etc are you using?

**Reply by adarsh2798** [reply]
Hi
Thanks for the reply. I am using python 3.8 and installed cflib from this page https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/installation/install/#install-cflib-from-source

**Reply by gemenerik** [reply]
What OS are you running?

**Reply by adarsh2798** [reply]
I am using Windows 11.
Main problem is If i load a crazyradio with more than 2 crazyflies The swarm class is not working as it is supposed to. It starts parallel threads for crazyflies connected to each radio one by one. So If I am using 5 crazyflies and 2 radios with distribution of 3+2, it first starts the 3 crazyfleis then after few seconds other 2 start. There is a considerable lag/delay.

But if I use 3 crazyradios and 6 crazyflies , 2 flies to one radio All 6 run parallelly.

**Reply by gemenerik** [reply]
I cannot reproduce it right now, but that sounds like a bug. It might have to do with iffy Python "multi"threading. Can you perhaps open an issue in [cflib](https://github.com/bitcraze/crazyflie-lib-python)? We will look at it. 

In the meantime consider playing around with [crazywarm2](https://github.com/IMRCLab/crazyswarm2)
